relay-server/src/queues/mod.rs

```diff
@@ -0,0 +1,51 @@
+use std::marker::PhantomData;
+use std::sync::{Arc, Mutex, MutexGuard};
+
+pub mod mem;
+
+// TODO: docs on all public types and functions.
+
+/// Combine multiple logical queues into a single structure so we can efficiently handle 10.000s of queues.
+pub trait MultiQueue<K, T> {
+    fn push_back(&mut self, queue_id: &K, item: T);
+    fn pop_front(&mut self, queue_id: &K) -> Option<T>;
+    fn len(&self, queue_id: &K) -> usize;
+}
+
+/// Behaves like a queue, but defers actual work to a MultiQueue backend.
+#[derive(Debug)]
+pub struct QueueView<K, T, Q: MultiQueue<K, T>> {
+    backend: Arc<Mutex<Q>>,
```

Comment:
Is `Arc<Mutex>` the way we want to synchronize, or should we use some kind of two-way channel instead?